# sc-guidelines
This repository contains the code and resources for the paper titled "Single-cell differential expression analysis between conditions within nested settings".
## Overview
In this study, we evaluated various methods for differential expression analysis in single-cell transcriptomics, addressing challenges related to pseudoreplication.
The tested methods include [DESeq2](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8), [MAST](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0844-5), [DREAM](https://academic.oup.com/bioinformatics/article/37/2/192/5878955), [scVI](https://www.nature.com/articles/s41592-018-0229-2), Permutation Test, and [distinct](https://www.biorxiv.org/content/10.1101/2020.11.24.394213v1). 
Additionally, we adapted Hierarchical Bootstrapping for single-cell data analysis. 
Our focus was on benchmarking various scenarios and providing method recommendations for further differential expression analysis studies"
